### PR TITLE
enable graph trainer + mxfp8 composability

### PR DESCRIPTION
**Pull request by vkuzo**
Summary:

1. add a config for graph trainer llama3 + mxfp8
2. add a config for graph trainer deepseek debug + mxfp8
3. make 2 actually call mxfp8 kernels (instead of silently falling back to bfloat16) by removing torchao's tensor subclass weight wrapper and calling the underlying `torch.autograd.Function` with the mxfp8 logic directly. This is needed because `GraphTrainer` uses SimpleFSDP, that uses DTensorized weight, and that does not compose with torchao's previous usage of `__torch_function__`.

An explanation of 3:

```code
# before
* `foo.bar.expert_weight` is a tensor subclass with a `__torch_function__` override that calls the mxfp8 logic
* the parent calls `torch._grouped_mm`, and we dispatch to the mxfp8 logic via the subclass override above

# after
* `foo.bar.expert_weight` is a regular parameter
* the parent calls `self._grouped_mm`
* model configuration controls whether `self._grouped_mm` dispatches to the default `torch._grouped_mm` or an override.  The mxfp8 modeling classes set the override to call the mxfp8 logic.
```

Same result in the end, but more direct and no longer depends on `__torch_function__`.

Test Plan:

```
// on 1 B200 GPU

// graph trainer + dense model + mxfp8
// verify the chrome trace contains mxfp8 casts + gemm
NGPU=1 MODULE=graph_trainer.llama3 CONFIG=graph_trainer_llama3_8b_mxfp8 ./run_train.sh

// graph trainer + moe debug model + mxfp8
// verify the chrome trace contains mxfp8 casts + gemm
NGPU=1 MODULE=graph_trainer.deepseek_v3 CONFIG=graph_trainer_deepseek_v3_debugmodel_mxfp8 ./run_train.sh

// EP 2
NGPU=2 MODULE=graph_trainer.deepseek_v3 CONFIG=graph_trainer_deepseek_v3_debugmodel_mxfp8 ./run_train.sh --parallelism.expert_parallel_degree=2
...
 step:  1  loss:  8.16138  grad_norm:  3.6610  memory:  2.98GiB(1.67%)  tps: 414  tflops: 0.23  mfu: N/A
 Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
 step:  2  loss:  6.18522  grad_norm:  4.6098  memory:  3.07GiB(1.72%)  tps: 17,394  tflops: 9.44  mfu: N/A
 step:  3  loss:  4.88102  grad_norm:  2.8651  memory:  3.07GiB(1.72%)  tps: 109,666  tflops: 59.54  mfu: N/A
 step:  4  loss:  4.76242  grad_norm:  2.6567  memory:  3.07GiB(1.72%)  tps: 131,115  tflops: 71.19  mfu: N/A
 step:  5  loss:  4.42710  grad_norm:  2.4591  memory:  3.07GiB(1.72%)  tps: 135,408  tflops: 73.52  mfu: N/A
 step:  6  loss:  4.20295  grad_norm:  2.0352  memory:  3.07GiB(1.72%)  tps: 122,923  tflops: 66.74  mfu: N/A
 step:  7  loss:  4.10372  grad_norm:  1.9203  memory:  3.09GiB(1.73%)  tps: 123,222  tflops: 66.90  mfu: N/A
 step:  8  loss:  4.01469  grad_norm:  2.0419  memory:  3.09GiB(1.73%)  tps: 135,611  tflops: 73.63  mfu: N/A
 step:  9  loss:  4.26402  grad_norm:  1.7035  memory:  3.09GiB(1.73%)  tps: 126,333  tflops: 68.59  mfu: N/A
 step: 10  loss:  3.89563  grad_norm:  1.5778  memory:  3.09GiB(1.73%)  tps: 100,589  tflops: 54.61  mfu: N/A

// tlparse of above: https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpZnVSv0/rank_0/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000


// verify eager mxfp8 did not change
> python3 scripts/loss_compare.py HEAD~1 HEAD --baseline-module=deepseek_v3 --baseline-config=deepseek_v3_debugmodel_mxfp8 --baseline-ngpus=1 --test-ngpus=1 --steps=100 --assert-equal --no-seed-checkpoint
...
[LOSS_COMPARE] 98      3.16805100440979    3.16805100440979   0.000000
[LOSS_COMPARE] 99      2.819762706756592    2.819762706756592   0.000000
[LOSS_COMPARE] 100     2.7529165744781494    2.7529165744781494   0.000000
[LOSS_COMPARE]
[LOSS_COMPARE] Summary statistics:
[LOSS_COMPARE] Average baseline loss:  3.1106575202941893
[LOSS_COMPARE] Average test loss: 3.1106575202941893
[LOSS_COMPARE] Average difference:     0.000000


// 8 GPU test: https://www.internalfb.com/phabricator/paste/view/P2422670858
// more single GPU tests: https://www.internalfb.com/intern/paste/P2422673885/
```

Verified the traces manually to see mxfp8 MoE kernels being called in GraphTrainer runs

### CODE REVIEW

**Comment by vkuzo** [comment]
ok, this approach works for dense but not moe. Reasoning, current torchao mxp8 moe training integration uses `__torch_function__`, and GraphTrainer does not trace `__torch_function__`.  My preferred fix would be to just use `torch.autograd.Function` for quantizing a `grouped_mm` call from torchtitan.

clauding ^: https://gist.github.com/vkuzo/dc0ac749cb778ea27707e078357f6899

File: test.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
```
Comment (vkuzo) [inline test.sh]: this file is not for land

File: test_single_gpu.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
```
Comment (vkuzo) [inline test_single_gpu.sh]: this file is not for land

**Comment by ezyang** [comment]
@SherlockNoMad I don't understand why the torch function subclass wouldn't work with non-strict tracer; in fact, it's something that I'd expect to work better than the Dynamo.

**Comment by ezyang** [comment]
@tianyu-l on your opinion on adding the `_grouped_mm` overrideable method as the way of overriding it. This is an aesthetic thing, it doesn't seem like a big deal to me but maybe you care.

**Comment by ezyang** [comment]
Show evidence that the loss is bitwise equivalent before and after this change?

**Comment by vkuzo** [comment]
> Show evidence that the loss is bitwise equivalent before and after this change?

for bf16, right? There is no baseline for GraphTrainer + mxfp8 since this PR unbreaks it.

File: torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py [hunk 3]
```diff
@@ -29,6 +31,29 @@ def graph_trainer_deepseek_v3_debugmodel() -> GraphTrainer.Config:
     return config
 
 
+def graph_trainer_deepseek_v3_debugmodel_mxfp8() -> GraphTrainer.Config:
```
Comment (ezyang) [inline torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py:37]: All of the `mxfp8` variants seem pretty boilerplatey--not sure if there is a way around it though.

Comment (ezyang) [reply torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py:37]: When I think about what low prec research on torchtitan might look at, the boilerplate seems increasingly bad, because there are bunch of different spots where I may want to swap between high and low precision; this `_mxfp8` config is ONLY the gmm (now granted, torchtitan popularized this particular recipe, so it's ok for it to be deified in some way), but there are plenty of other ways to "do mxfp8".

File: torchtitan/models/common/token_dispatcher.py [hunk 4]
```diff
@@ -673,9 +673,11 @@ class TorchAOTokenDispatcher(AllToAllTokenDispatcher):
...
     quantized grouped GEMM kernels (e.g. 16 for FP8, 32 for MXFP8).
 
-    Requires EP to be enabled (ep_mesh must be set). Raises ValueError
-    if ep_mesh is None, since quantized grouped GEMMs need padded token
-    groups which are only produced by the EP permute_and_pad path.
+    Works with EP enabled (all-to-all dispatch + padded permute) and with
+    EP=1 (``ep_mesh is None``), where it skips the all-to-all and only applies
+    the local padded permute. The padding is what the quantized grouped GEMM
+    needs; the all-to-all is orthogonal, so EP=1 is supported for single-GPU
+    debugging / numerics by running ``permute_and_pad`` with ``ep_degree=1``.
```
Comment (ezyang) [inline torchtitan/models/common/token_dispatcher.py:680]: The local support is new I guess?

**Comment by vkuzo** [comment]
> I don't understand why the torch function subclass wouldn't work with non-strict tracer; in fact, it's something that I'd expect to work better than the Dynamo.

btw, we want to do this regardless of GraphTrainer limitations, for more direct code.  The feedback we got overwhelmingly is that subclasses used to override ops are confusing / unintuitive.

**Comment by ezyang** [comment]
> for bf16, right? There is no baseline for GraphTrainer + mxfp8 since this PR unbreaks it.

For eager mxfp8 (since it seems like this used to exist).

File: torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py (hunk 3 above)
Comment (ezyang) [reply torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py:37]: Shouldn't you add tests for these configs? I'm not sure I want to spend GB200x2 CI hours on it, but I would certainly like the no-EP single GPU config to be tested with deterministic loss check.

**Review by ezyang — APPROVED** [review]

**Comment by vkuzo** [comment]
> For eager mxfp8 (since it seems like this used to exist).

makes sense, will add to the PR summary shortly

**Comment by vkuzo** [comment]
> For eager mxfp8 (since it seems like this used to exist).

This PR also adds the eager config for deepseek mxfp8 as well as mxfp8 + EP=1 support, I'll split that to a separate PR to make it easier to track numerics changes for the graph trainer related changes.

File: torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py [hunk 5]
```diff
@@ -29,6 +31,29 @@ def graph_trainer_deepseek_v3_debugmodel() -> GraphTrainer.Config:
...
+    base = deepseek_v3_debugmodel()
+    # Quantize the MoE expert grouped GEMMs to MXFP8 on the base model config
+    # before wrapping in the graph_trainer config. to_graph_trainer_config copies
+    # the (now MXFP8-swapped) model fields into the graph model class and swaps in
+    # graph_trainer's parallelize_fn. graph_trainer always compiles the model, so
+    # the converter's compile requirement is satisfied. pad_multiple=128 is
+    # required by the CuTeDSL quantization kernel on sm_100 (e.g. B200); the
+    # default of 32 only suffices on older architectures.
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py]: wondering how did you do testing? 
If you have verified manually, graph trainer also have some e2e unit test, and we should have 2 test cases that uses the mxfp8 configs.

File: torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py [hunk 6]
```diff
@@ -29,6 +31,29 @@ def graph_trainer_deepseek_v3_debugmodel() -> GraphTrainer.Config:
     return config
 
 
+def graph_trainer_deepseek_v3_debugmodel_mxfp8() -> GraphTrainer.Config:
+    base = deepseek_v3_debugmodel()
+    # Quantize the MoE expert grouped GEMMs to MXFP8 on the base model config
+    # before wrapping in the graph_trainer config. to_graph_trainer_config copies
+    # the (now MXFP8-swapped) model fields into the graph model class and swaps in
+    # graph_trainer's parallelize_fn. graph_trainer always compiles the model, so
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py]: >graph_trainer always compiles the model

That's not the case... GraphTrainer always trace the model, but only regional compiles flex_attn. 
I think we also need to mark the quantized region with regional_inductor?

File: torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py [hunk 7]
```diff
@@ -29,6 +31,29 @@ def graph_trainer_deepseek_v3_debugmodel() -> GraphTrainer.Config:
     return config
 
 
+def graph_trainer_deepseek_v3_debugmodel_mxfp8() -> GraphTrainer.Config:
+    base = deepseek_v3_debugmodel()
+    # Quantize the MoE expert grouped GEMMs to MXFP8 on the base model config
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py]: what about the other gemms not in the MoE regions?

**Comment by SherlockNoMad** [comment]
can you share a tlprase link to the run? 

I am wondering how it work without compiling?

**Comment by SherlockNoMad** [comment]
> @SherlockNoMad I don't understand why the torch function subclass wouldn't work with non-strict tracer; in fact, it's something that I'd expect to work better than the Dynamo.

The original problem was not tracer not working with torch_function subclass. 
The problem was SimpleFSDP is using DTensorized weight, which as composability problem with torch_function subclass.

**Comment by vkuzo** [comment]
I split out mxfp8+eager baseline and mxfp8+ep=1 baselines to this stack: https://github.com/pytorch/torchtitan/pull/3935

will rebase this PR on top once they land

File: torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py (hunk 5 above)
Comment (vkuzo) [reply torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py]: yeah, let's do that after https://github.com/pytorch/torchtitan/pull/3935 lands, we need a clean baseline to add tests

File: torchtitan/experiments/graph_trainer/llama3/config_registry.py [hunk 8]
```diff
@@ -73,6 +75,20 @@ def graph_trainer_llama3_8b() -> GraphTrainer.Config:
...
 
 
+def graph_trainer_llama3_8b_mxfp8() -> GraphTrainer.Config:
+    base = llama3_8b()
+    # Swap dense Linear layers for MXFP8Linear before wrapping in the
+    # graph_trainer config. graph_trainer always compiles the model, so the
+    # MXFP8 converter's compile requirement is satisfied.
+    base.model_spec = llama3_model_registry(
+        "8B",
+        converters=[MXFP8LinearConverter.Config(model_compile_enabled=True)],
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/llama3/config_registry.py:85]: I see you did the fix for grouped_mm, but did't see the fix for linear. 

Is MXFP8LinearConverter not using torch_function?

File: torchtitan/models/gpt_oss/moe.py
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/moe.py]:
```suggestion
class GptOssGroupedExperts(GroupedExperts):
```
can we do this and remove the `self._grouped_mm` overwrite in this class?

File: torchtitan/models/common/moe.py [hunk 9]
```diff
@@ -94,21 +94,33 @@ def _experts_forward(
...
+            offsets_E,
         )
-        return torch._grouped_mm(
-            h_RF, w2_EDF.bfloat16().transpose(-2, -1), offs=offsets_E
+        return self._grouped_mm(
+            h_RF, w2_EDF.bfloat16().transpose(-2, -1), offsets_E
         ).type_as(x_RD)
 
+    def _grouped_mm(
+        self, A: torch.Tensor, B_t: torch.Tensor, offs: torch.Tensor
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]:
```suggestion
        self, *, A: torch.Tensor, B_t: torch.Tensor, offs: torch.Tensor
```
so that they are all kwargs, for readability

File: torchtitan/experiments/graph_trainer/llama3/config_registry.py (hunk 8 above)
Comment (tianyu-l) [reply torchtitan/experiments/graph_trainer/llama3/config_registry.py:85]: seems so https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/quantization/mx.py#L39

**Review by tianyu-l — APPROVED** [review]
please address comments before merging

**Review by SherlockNoMad — APPROVED** [review]

File: torchtitan/experiments/graph_trainer/llama3/config_registry.py (hunk 8 above)
Comment (vkuzo) [reply torchtitan/experiments/graph_trainer/llama3/config_registry.py:85]: > Is MXFP8LinearConverter not using torch_function?

yes, it is not

File: torchtitan/models/common/moe.py (hunk 9 above)
Comment (vkuzo) [reply torchtitan/models/common/moe.py]: done

**Comment by vkuzo** [comment]
> Show evidence that the loss is bitwise equivalent before and after this change?
> For eager mxfp8 (since it seems like this used to exist).

yes, just updated the test plan with evidence of this for mxfp8 eager, with NGPU1

**Comment by vkuzo** [comment]
CI failure looks unrelated, landing
